rts/src/gc.c

```diff
@@ -0,0 +1,234 @@
+#include <assert.h>
+#include <stdlib.h>
+
+#include "tim.h"
+#include "mem.h"
+#include "dump.h"
+#include "gc.h"
+
+#ifndef NO_GC
+
+/*****************/
+/* Mark handling */
+/*****************/
+Mark gc_mark=1;
+
+void mark_refresh() {
+  gc_mark = (gc_mark + 1) % 100;
+  if (gc_mark <= 0) {gc_mark = 1;}
+}
+
+/**********************/
+/* location recording */
+/**********************/
+
+gc_data data_locations;
+
+void add_location(void* location) {
+  gc_list new_location = rts_malloc(sizeof(struct gc_list));
+  new_location->location = location;
+  new_location->next = data_locations.locations;
+  data_locations.locations = new_location;
+  data_locations.size++;
+}
+
+/**********************/
+/* Deallocation (free)*/
+/**********************/
+
+void free_location(void* ptr) {
+  Mark* location = (Mark*) ptr;
+  location--;
+  rts_free(location);
+}
+
+/***********/
+/* Marking */
+/***********/
+
+// Prototypes
+void mark_frame(frame_t frame);
+
+void mark_location(void* ptr) {
+  if (ptr == NULL) { return; }
+  Mark* location = (Mark*) ptr;
+  location--;
+  *location = gc_mark;
+}
+
+int is_marked_location(void* ptr) {
+  Mark* location = (Mark*) ptr;
+  if (location == NULL) { return 1; }
+  location--;
+  return *location == gc_mark;
+}
+
+int is_regular_closure(closure_t* closure) {
+  return closure->code != *tim_value_code &&
+         closure->code != *tim_nil_code;
+}
```

Comment:
Your scientists were so preoccupied with whether or not they could, that they didn’t stop to think if they should.

Reply:
Una busqueda rapida en google mostro que StackOverflow piensa que es safe 🤣